Esp8266/Esp8266.ino

```diff
@@ -1,86 +1,116 @@
-#include "event.pb.h"
-
-#include <pb_common.h>
-#include <pb.h>
-#include <pb_encode.h>
-#include <pb_decode.h>
+#include "pb_common.h"
+#include "pb.h"
+#include "pb_encode.h"
+#include "pb_decode.h"
+#include "Esp8266_Interface.pb.h"
 
 #include <ESP8266WiFi.h>
 
-const char* ssid     = "onhub";
-const char* password = "ekchotaghar";
-const char* addr     = "10.0.1.181";
-const uint16_t port  = 10100;
+const char* ssid     = "SSID";
+const char* pass     = "PASS";
+const uint16_t port  = 42069;
 
+WiFiServer server(port);
 WiFiClient client;
 
-void sendStat(pb_Event e);
+struct InputData {
+  static const unsigned buffer_length = esp8266_Esp8266_Command_size;
+  unsigned char buffer[buffer_length];
+  pb_istream_t input_stream;
+  esp8266_Esp8266_Command command;
 
-// setup WIFI and sensor
-void setup() {
-  pinMode(LED_BUILTIN, OUTPUT);
-  Serial.begin(115200);
-  delay(10);
+  InputData() { refresh(); };
 
-  Serial.println();
-  Serial.print("Setting up WIFI for SSID ");
-  Serial.println(ssid);
+  void refresh() {
+    input_stream = pb_istream_from_buffer(buffer, buffer_length);
+    command = esp8266_Esp8266_Command_init_zero;
+  }
+
+  void readAndDecode() {
+    refresh();
+    client.read(buffer, client.available());
+    pb_decode(&input_stream, esp8266_Esp8266_Command_fields, &command);
+  }
+} input;
 
-//  WiFi.mode(WIFI_STA);
-  WiFi.begin(ssid, password);
+struct OutputData {
+  static const unsigned buffer_length = esp8266_Esp8266_Metrics_size;
+  unsigned char buffer[buffer_length];
+  pb_ostream_t output_stream;
+  esp8266_Esp8266_Metrics metrics;
 
-  while (WiFi.status() != WL_CONNECTED) {
-    Serial.println("WIFI connection failed, reconnecting...");
-    delay(500);
+  OutputData() { refresh(); }
+  void refresh() {
+    output_stream = pb_ostream_from_buffer(buffer, buffer_length);
   }
-  
-  Serial.print("\nWiFi connected, ");
-  Serial.print("IP address: ");
-  Serial.println(WiFi.localIP());
+
+  void encodeAndWrite() {
+    refresh();
+    pb_encode(&output_stream, esp8266_Esp8266_Metrics_fields, &metrics);
+    client.write(buffer, output_stream.bytes_written);
+  }
+} output;
+
+void setupWiFi();
+void printConnection(const char * name, IPAddress address, int port);
+
+void handleData();
+
+// setup WIFI and sensor
+void setup() {
+  Serial.begin(9600);
+  setupWiFi();
 }
 
 
 void loop() {
-  digitalWrite(LED_BUILTIN, LOW);
-  Serial.print("connecting to ");
-  Serial.println(addr);
-  client.connect(addr, port);
-
-  if (!client.connect(addr, port)) {
-    Serial.println("connection failed");
-    Serial.println("wait 5 sec to reconnect...");
-    delay(5000);
-    return;
+  client = server.available();
+  if (client) {
+    printConnection("controller", client.remoteIP(), client.remotePort());
+    client.setNoDelay(true);
```

Comment:
What does this do?

Reply:
Every loop(), it waits for a connection. If available() ends, it checks if there really was a connection and then prints that there was one. Then there is another loop for handling that specific client. Once that client disconnects, it prints and goes back to the top of loop() to handle ANOTHER client.